File: packages/nerdster_common/lib/trust_logic.dart
import 'dart:collection';
import 'package:oneofus_common/statement.dart';
import 'package:oneofus_common/trust_statement.dart';
import 'package:oneofus_common/keys.dart';
import 'package:nerdster_common/trust_graph.dart';

final Map<String, String> pathsReq = {
  'standard': '1-1-2',
  'strict': '1-2-2-3',
  'permissive': '1',
};

typedef PathRequirement = int Function(int distance);

List<List<IdentityKey>> _findNodeDisjointPaths(
    IdentityKey root, IdentityKey target, Map<IdentityKey, Set<IdentityKey>> graph, int limit) {
  final List<List<IdentityKey>> paths = [];
  final Set<IdentityKey> excludedNodes = {};
  final Set<String> usedPathStrings = {};

  while (paths.length < limit) {
    final path = _findShortestPath(root, target, graph, excludedNodes);
    if (path == null) break;

    final pathString = path.map((k) => k.value).join('->');
    if (usedPathStrings.contains(pathString)) break;

    paths.add(path);
    usedPathStrings.add(pathString);

    for (int i = 1; i < path.length - 1; i++) {
      excludedNodes.add(path[i]);
    }
  }
  return paths;
}

List<IdentityKey>? _findShortestPath(IdentityKey start, IdentityKey end,
    Map<IdentityKey, Set<IdentityKey>> graph, Set<IdentityKey> excluded) {
  final Queue<List<IdentityKey>> queue = Queue<List<IdentityKey>>();
  queue.add([start]);
  final Set<IdentityKey> visited = {start, ...excluded};

  while (queue.isNotEmpty) {
    final List<IdentityKey> path = queue.removeFirst();
    final IdentityKey node = path.last;

    if (node == end) return path;

    final Set<IdentityKey>? neighbors = graph[node];
    if (neighbors != null) {
      for (final IdentityKey neighbor in neighbors) {
        if (!visited.contains(neighbor)) {
          visited.add(neighbor);
          final List<IdentityKey> newPath = List<IdentityKey>.from(path)..add(neighbor);
          queue.add(newPath);
        }
      }
    }
  }
  return null;
}

TrustGraph reduceTrustGraph(
  TrustGraph current,
  Map<IdentityKey, List<TrustStatement>> byIssuer, {
  PathRequirement? pathRequirement,
  int maxDegrees = 6,
}) {
  Statement.validateOrderTypess(byIssuer.values);

  final Map<IdentityKey, int> distances = {current.pov: 0};
  final List<IdentityKey> orderedKeys = [current.pov];
  final Map<IdentityKey, IdentityKey> equivalent2canonical = {};
  final Set<IdentityKey> blocked = {};
  final Map<IdentityKey, List<List<IdentityKey>>> paths = {};
  final List<TrustNotification> notifications = [];
  final Map<IdentityKey, List<TrustStatement>> edges = {};
  final Map<IdentityKey, Set<IdentityKey>> trustedBy = {};
  final Map<IdentityKey, Set<IdentityKey>> graphForPathfinding = {};
  final Set<IdentityKey> visited = {current.pov};

  IdentityKey resolveCanonical(IdentityKey token) {
    IdentityKey currentKey = token;
    final Set<IdentityKey> seen = {token};
    while (equivalent2canonical.containsKey(currentKey)) {
      currentKey = equivalent2canonical[currentKey]!;
      if (seen.contains(currentKey)) break;
      seen.add(currentKey);
    }
    return currentKey;
  }

  final PathRequirement req = pathRequirement ?? (d) => 1;

  Set<IdentityKey> currentLayer = {current.pov};

  for (int dist = 0; dist < maxDegrees && currentLayer.isNotEmpty; dist++) {
    final Set<IdentityKey> nextLayer = <IdentityKey>{};

    for (final IdentityKey issuer in currentLayer) {
      List<TrustStatement> statements = byIssuer[issuer] ?? [];
      final Set<IdentityKey> decided = <IdentityKey>{};

      for (final TrustStatement s
          in statements.where((TrustStatement s) => s.verb == TrustVerb.block)) {
        final IdentityKey subject = s.subjectAsIdentity;
        if (decided.contains(subject)) continue;
        decided.add(subject);

        if (subject == current.pov) {
          notifications.add(TrustNotification(
            reason: "Attempt to block your key.",
            rejectedStatement: s,
            isConflict: true,
          ));
          continue;
        }

        if (distances.containsKey(subject) && distances[subject]! <= dist) {
          notifications.add(TrustNotification(
            reason: "Attempt to block trusted key by ${issuer.value}",
            rejectedStatement: s,
            isConflict: true,
          ));
        } else {
          blocked.add(subject);
        }
      }
    }

    for (final IdentityKey issuer in currentLayer) {
      List<TrustStatement> statements =
          equivalent2canonical.containsKey(issuer) ? [] : (byIssuer[issuer] ?? []);

      edges[issuer] = statements.where((TrustStatement s) {
        if (s.verb == TrustVerb.clear) return false;
        if (s.verb != TrustVerb.replace && s.verb != TrustVerb.delegate && s.revokeAt != null)
          return false;
        return true;
      }).toList();

      final Set<IdentityKey> decided = <IdentityKey>{};
      for (final TrustStatement s
          in statements.where((TrustStatement s) => s.verb == TrustVerb.replace)) {
        final IdentityKey oldKey = s.subjectAsIdentity;
        if (decided.contains(oldKey)) continue;
        decided.add(oldKey);
        assert(s.revokeAt == kSinceAlways || s.revokeAt == null,
            'replace with revokeAt other than <since always> is not supported: ${s.revokeAt}');

        if (oldKey == current.pov) {
          notifications.add(TrustNotification(
            reason: "Attempt to replace your key.",
            rejectedStatement: s,
            isConflict: true,
          ));
          continue;
        }

        if (blocked.contains(oldKey)) {
          notifications.add(TrustNotification(
            reason: "Blocked key ${oldKey.value} is being replaced by ${issuer.value}",
            rejectedStatement: s,
            isConflict: false,
          ));
          continue;
        }

        if (distances.containsKey(oldKey) && distances[oldKey]! < dist) {
          if (!equivalent2canonical.containsKey(oldKey)) {
            equivalent2canonical[oldKey] = issuer;
          }
          notifications.add(TrustNotification(
            reason:
                "Trusted key ${oldKey.value} is being replaced by ${issuer.value} (Replacement constraint ignored due to distance)",
            rejectedStatement: s,
            isConflict: false,
          ));
          continue;
        }

        if (equivalent2canonical.containsKey(oldKey)) {
          final IdentityKey? existingNewKey = equivalent2canonical[oldKey];
          if (existingNewKey != issuer) {
            notifications.add(TrustNotification(
              reason:
                  "Key ${oldKey.value} replaced by both ${existingNewKey!.value} and ${issuer.value}",
              rejectedStatement: s,
              isConflict: true,
            ));
            continue;
          }
        }

        if (distances.containsKey(oldKey)) {
          notifications.add(TrustNotification(
            reason: "Trusted key ${oldKey.value} is being replaced by ${issuer.value}",
            rejectedStatement: s,
            isConflict: false,
          ));
        }

        equivalent2canonical[oldKey] = issuer;
        graphForPathfinding.putIfAbsent(issuer, () => {}).add(oldKey);

        if (!visited.contains(oldKey)) {
          visited.add(oldKey);
          distances[oldKey] = dist + 1;
          orderedKeys.add(oldKey);
          nextLayer.add(oldKey);
        }
      }
    }

    for (final IdentityKey issuer in currentLayer) {
      final List<TrustStatement> statements =
          equivalent2canonical.containsKey(issuer) ? [] : (edges[issuer] ?? []);

      final Set<IdentityKey> decided = <IdentityKey>{};
      for (final TrustStatement s
          in statements.where((TrustStatement s) => s.verb == TrustVerb.trust)) {
        final IdentityKey subject = s.subjectAsIdentity;
        if (decided.contains(subject)) continue;
        decided.add(subject);

        if (blocked.contains(subject)) {
          notifications.add(TrustNotification(
            reason: "Attempt to trust blocked key by ${issuer.value}",
            rejectedStatement: s,
            isConflict: true,
          ));
          continue;
        }

        IdentityKey effectiveSubject = resolveCanonical(subject);

        if (blocked.contains(effectiveSubject)) continue;

        trustedBy.putIfAbsent(effectiveSubject, () => {}).add(issuer);

        final int requiredPaths = req(dist + 1);

        for (final IdentityKey i in trustedBy[effectiveSubject]!) {
          graphForPathfinding.putIfAbsent(i, () => {}).add(effectiveSubject);
        }

        final List<List<IdentityKey>> foundPaths = _findNodeDisjointPaths(
            current.pov, effectiveSubject, graphForPathfinding, requiredPaths);
        if (foundPaths.length >= requiredPaths) {
          paths[effectiveSubject] = foundPaths;
          if (!visited.contains(effectiveSubject)) {
            visited.add(effectiveSubject);
            distances[effectiveSubject] = dist + 1;
            orderedKeys.add(effectiveSubject);
            nextLayer.add(effectiveSubject);
          }
          if (effectiveSubject != subject) {
            distances[subject] = dist + 1;
            if (!visited.contains(subject)) {
              visited.add(subject);
              orderedKeys.add(subject);
              nextLayer.add(subject);
            }
          }
        }
      }
    }

    currentLayer = nextLayer;
  }

  final Map<String, TrustNotification> uniqueNotifications = {};
  for (final TrustNotification n in notifications) {
    final String key = "${n.subject.value}:${n.reason}";
    if (!uniqueNotifications.containsKey(key)) {
      uniqueNotifications[key] = n;
    }
  }

  return TrustGraph(
    pov: current.pov,
    distances: distances,
    orderedKeys: orderedKeys,
    equivalent2canonical: equivalent2canonical,
    blocked: blocked,
    paths: paths,
    notifications: uniqueNotifications.values.toList(),
    edges: edges,
  );
}

File: functions/trust_logic.js
/**
 * trust_logic.js — JavaScript port of trust_logic.dart
 *
 * Implements the same greedy BFS as the Dart Nerdster.
 * Source interface: { fetch(fetchMap: {[token]: null}) => Promise<{[token]: statement[]}> }
 */

const { getToken } = require('./jsonish_util');

const kSinceAlways = '<since always>';
const kDefaultMaxDegrees = 6;
const kTrustVerbs = ['trust', 'block', 'replace', 'clear', 'delegate'];

function permissivePathRequirement(_distance) {
  return 1;
}

function defaultPathRequirement(distance) {
  if (distance <= 3) return 1;
  if (distance <= 4) return 2;
  return 3;
}

function strictPathRequirement(distance) {
  if (distance <= 2) return 1;
  if (distance <= 3) return 2;
  return 3;
}

// ---------------------------------------------------------------------------
// Statement parsing
// ---------------------------------------------------------------------------

function getVerbSubject(s) {
  for (const verb of kTrustVerbs) {
    if (s[verb] != null) return { verb, subject: s[verb] };
  }
  return null;
}

async function parseStatements(statements) {
  return Promise.all(statements.map(async (s) => {
    const iToken = await getToken(s.I);
    const vs = getVerbSubject(s);
    const subjectToken = vs ? await getToken(vs.subject) : null;
    return {
      iToken,
      subjectToken,
      subjectPubKey: vs?.subject ?? null,
      verb: vs?.verb ?? null,
      revokeAt: s.with?.revokeAt ?? null,
      endpoint: s.with?.endpoint?.url ?? null,
      moniker: s.with?.moniker ?? null,
      raw: s,
    };
  }));
}

// ---------------------------------------------------------------------------
// Path-finding (node-disjoint BFS)
// ---------------------------------------------------------------------------

function _findShortestPath(start, end, graph, excluded) {
  const queue = [[start]];
  const visited = new Set([start, ...excluded]);

  while (queue.length > 0) {
    const path = queue.shift();
    const node = path[path.length - 1];
    if (node === end) return path;
    for (const neighbor of (graph.get(node) || [])) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push([...path, neighbor]);
      }
    }
  }
  return null;
}

function _findNodeDisjointPaths(root, target, graph, limit) {
  const paths = [];
  const excludedNodes = new Set();
  const usedPathStrings = new Set();

  while (paths.length < limit) {
    const path = _findShortestPath(root, target, graph, excludedNodes);
    if (!path) break;
    const pathString = path.join('->');
    if (usedPathStrings.has(pathString)) break;
    paths.push(path);
    usedPathStrings.add(pathString);
    for (let i = 1; i < path.length - 1; i++) excludedNodes.add(path[i]);
  }
  return paths;
}

// ---------------------------------------------------------------------------
// Core reducer — always rebuilds from scratch using all accumulated statements.
// Only current.pov is used from the current graph; everything else is recomputed.
// ---------------------------------------------------------------------------

async function reduceTrustGraph(pov, byIssuer, { pathRequirement, maxDegrees = kDefaultMaxDegrees, fedRegistry } = {}) {
  const req = pathRequirement || defaultPathRequirement;

  // Pre-parse all statements so BFS iteration is synchronous
  const parsed = new Map(); // issuerToken -> [parsedStatement]
  for (const [token, statements] of byIssuer) {
    parsed.set(token, await parseStatements(statements));
  }

  // Register any foreign endpoints discovered in this round of statements
  if (fedRegistry) {
    for (const stmtList of parsed.values()) {
      for (const s of stmtList) {
        if (s.endpoint && s.subjectToken) fedRegistry.set(s.subjectToken, s.endpoint);
      }
    }
  }

  const distances = new Map([[pov, 0]]);
  const orderedKeys = [pov];
  const equivalent2canonical = new Map();
  const blocked = new Set();
  const paths = new Map();
  const notifications = [];
  const edges = new Map();
  const trustedBy = new Map();
  const graphForPathfinding = new Map();
  const visited = new Set([pov]);
  const monikers = new Map(); // canonicalToken → string[] (in BFS order, first = best)
  const pubKeys = new Map(); // token → pubKeyJson

  // Collect pubKeys from self-signed statements (delegate statements signed by identity key).
  for (const [token, stmts] of parsed) {
    for (const s of stmts) {
      if (s.iToken === token && !pubKeys.has(token)) {
        pubKeys.set(token, s.raw.I);
        break;
      }
    }
  }

  function resolveCanonical(token) {
    let cur = token;
    const seen = new Set([token]);
    while (equivalent2canonical.has(cur)) {
      cur = equivalent2canonical.get(cur);
      if (seen.has(cur)) break;
      seen.add(cur);
    }
    return cur;
  }

  function addToGraph(from, to) {
    if (!graphForPathfinding.has(from)) graphForPathfinding.set(from, new Set());
    graphForPathfinding.get(from).add(to);
  }

  let currentLayer = new Set([pov]);

  for (let dist = 0; dist < maxDegrees && currentLayer.size > 0; dist++) {
    const nextLayer = new Set();

    // --- STAGE 1: BLOCKS ---
    for (const issuerToken of currentLayer) {
      const stmts = parsed.get(issuerToken) || [];
      const decided = new Set();
      for (const s of stmts) {
        if (s.verb !== 'block') continue;
        if (decided.has(s.subjectToken)) continue;
        decided.add(s.subjectToken);
        if (s.subjectToken === pov) {
          notifications.push({ reason: 'Attempt to block your key.', raw: s.raw, isConflict: true });
          continue;
        }
        if (distances.has(s.subjectToken) && distances.get(s.subjectToken) <= dist) {
          notifications.push({ reason: `Attempt to block trusted key by ${issuerToken}`, raw: s.raw, isConflict: true });
        } else {
          blocked.add(s.subjectToken);
        }
      }
    }

    // --- STAGE 2a: REPLACES ---
    for (const issuerToken of currentLayer) {
      let stmts = parsed.get(issuerToken) || [];
      if (equivalent2canonical.has(issuerToken)) stmts = [];

      const edgeStmts = stmts.filter(s => {
        if (s.verb === 'clear') return false;
        if (s.verb !== 'replace' && s.verb !== 'delegate' && s.revokeAt != null) return false;
        return true;
      });
      edges.set(issuerToken, edgeStmts);

      const decided = new Set();
      for (const s of stmts) {
        if (s.verb !== 'replace') continue;
        if (s.revokeAt !== kSinceAlways) {
          throw new Error(`replace with revokeAt other than <since always> is not supported: ${s.revokeAt}`);
        }
        const oldKey = s.subjectToken;
        if (decided.has(oldKey)) continue;
        decided.add(oldKey);

        if (oldKey === pov) {
          notifications.push({ reason: 'Attempt to replace your key.', raw: s.raw, isConflict: true });
          continue;
        }
        if (blocked.has(oldKey)) {
          notifications.push({ reason: `Blocked key ${oldKey} is being replaced by ${issuerToken}`, raw: s.raw, isConflict: false });
          continue;
        }
        if (distances.has(oldKey) && distances.get(oldKey) < dist) {
          if (!equivalent2canonical.has(oldKey)) equivalent2canonical.set(oldKey, issuerToken);
          notifications.push({ reason: `Trusted key ${oldKey} is being replaced by ${issuerToken} (Replacement constraint ignored due to distance)`, raw: s.raw, isConflict: false });
          continue;
        }
        if (equivalent2canonical.has(oldKey)) {
          const existing = equivalent2canonical.get(oldKey);
          if (existing !== issuerToken) {
            notifications.push({ reason: `Key ${oldKey} replaced by both ${existing} and ${issuerToken}`, raw: s.raw, isConflict: true });
            continue;
          }
        }
        if (distances.has(oldKey)) {
          notifications.push({ reason: `Trusted key ${oldKey} is being replaced by ${issuerToken}`, raw: s.raw, isConflict: false });
        }
        equivalent2canonical.set(oldKey, issuerToken);
        addToGraph(issuerToken, oldKey);
        if (!visited.has(oldKey)) {
          visited.add(oldKey);
          distances.set(oldKey, dist + 1);
          orderedKeys.push(oldKey);
          nextLayer.add(oldKey);
        }
      }
    }

    // --- STAGE 2b: TRUSTS ---
    for (const issuerToken of currentLayer) {
      let stmts = edges.get(issuerToken) || [];
      if (equivalent2canonical.has(issuerToken)) stmts = [];

      const decided = new Set();
      for (const s of stmts) {
        if (s.verb !== 'trust') continue;
        if (decided.has(s.subjectToken)) continue;
        decided.add(s.subjectToken);

        if (blocked.has(s.subjectToken)) {
          notifications.push({ reason: `Attempt to trust blocked key by ${issuerToken}`, raw: s.raw, isConflict: true });
          continue;
        }

        const effectiveSubject = resolveCanonical(s.subjectToken);
        if (blocked.has(effectiveSubject)) continue;

        if (!trustedBy.has(effectiveSubject)) trustedBy.set(effectiveSubject, new Set());
        trustedBy.get(effectiveSubject).add(issuerToken);

        const requiredPaths = req(dist + 1);
        const searchLimit = Math.max(requiredPaths, strictPathRequirement(dist + 1));
        for (const i of trustedBy.get(effectiveSubject)) addToGraph(i, effectiveSubject);

        const foundPaths = _findNodeDisjointPaths(pov, effectiveSubject, graphForPathfinding, searchLimit);
        if (foundPaths.length >= requiredPaths) {
          // Collect moniker (BFS order ensures first seen = from closest issuer).
          if (s.moniker) {
            if (!monikers.has(effectiveSubject)) monikers.set(effectiveSubject, []);
            const list = monikers.get(effectiveSubject);
            if (!list.includes(s.moniker)) list.push(s.moniker);
          }
          // Collect pubKey: use subject directly when it's the canonical token.
          if (s.subjectPubKey && s.subjectToken === effectiveSubject && !pubKeys.has(effectiveSubject)) {
            pubKeys.set(effectiveSubject, s.subjectPubKey);
          }
          paths.set(effectiveSubject, foundPaths);
          if (!visited.has(effectiveSubject)) {
            visited.add(effectiveSubject);
            distances.set(effectiveSubject, dist + 1);
            orderedKeys.push(effectiveSubject);
            nextLayer.add(effectiveSubject);
          }
          if (effectiveSubject !== s.subjectToken) {
            distances.set(s.subjectToken, dist + 1);
            if (!visited.has(s.subjectToken)) {
              visited.add(s.subjectToken);
              orderedKeys.push(s.subjectToken);
              nextLayer.add(s.subjectToken);
            }
          }
        }
      }
    }

    currentLayer = nextLayer;
  }

  // Deduplicate notifications
  const uniqueNotifications = new Map();
  for (const n of notifications) {
    const vs = getVerbSubject(n.raw);
    const subjectToken = vs ? await getToken(vs.subject) : await getToken(n.raw.I);
    const key = `${subjectToken}:${n.reason}`;
    if (!uniqueNotifications.has(key)) uniqueNotifications.set(key, n);
  }

  return {
    pov,
    distances,
    orderedKeys,
    equivalent2canonical,
    blocked,
    paths,
    notifications: [...uniqueNotifications.values()],
    monikers,
    pubKeys,
  };
}

module.exports = { reduceTrustGraph, permissivePathRequirement, defaultPathRequirement, strictPathRequirement, kSinceAlways, kDefaultMaxDegrees };
